x-pack/solutions/observability/plugins/observability/server/lib/rules/custom_threshold/custom_threshold_executor.ts

```diff
@@ -250,6 +251,8 @@ export const createCustomThresholdExecutor = ({
         );
 
         const groups = groupByKeysObjectMapping[group];
+        const grouping: Record<string, string> = {};
+        groups?.forEach((groupObj) => (grouping[groupObj.field] = groupObj.value));
```

Comment:
We need to move away from relying on alert instance Id to get group values as it can sometimes result in wrong group values. We are using `getFormattedGroupBy` function to get `groups` array which internally parses alert instance Id to get group values. Instead, we should generate group values from rule evaluation results (buckets). Related ticket: https://github.com/elastic/kibana/issues/220006.

Since we are adding a new field in alert document, I would suggest to generate group values from evaluation results instead of alert instance Id.

Reply:
Good point, thanks for bringing up this topic (glad that Patrick noticed this issue).
I fixed it mostly in this commit: [e80b47f](https://github.com/elastic/kibana/pull/219826/commits/e80b47f79426d398bc0287d560945ff50b1eb074) I wasn't able to remove some utilities as they are also used in the log threshold rule, but I added a todo to remove them when we change the logic in that rule.